### PR TITLE
implement KubernetesExecutor

### LINKED ISSUE #2: Extend Airflow operator by implementing KubernetesExecutor
As a user I want to have the option of running my airflow DAGs with the KubernetesExecutor, so that I have greater control over resource configuration (some settings can be defined per job) and usage (each job runs in its own pod which is created on-demand).

#### Implementation

- the airflow controller must define a pod template according to the specification details see [here](https://airflow.apache.org/docs/apache-airflow/stable/executor/kubernetes.html#configuration)
- this template must be mounted via e.g. PVC at the location defined by `AIRFLOW__KUBERNETES__POD_TEMPLATE_FILE` (see [configuration](https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#pod-template-file))
- if the airflow resource specifies KubernetesExecutor, then the scheduler recognises this and the KubernetesExecutor requests a worker pod from the Kubernetes API according the the template definition

#### Background/Context

Currently the airflow-operator implements the CeleryExecutor (Local- and SequentialExecutors are also supported but are not scalable) whereby webserver and scheduler pods interact with multiple (celery-)worker pods: celery reads job data from the external database and queues jobs via an external Redis instance. There are other executors available:

- `KubernetesExecutor`
  - each job is spun up in its own pod, which is then destroyed
  - no queue component is needed
  - accessing logs is more complicated
  - not sure if complex jobs can be distributed over multiple workers (as is the case with Celery)

The full list is here: https://airflow.apache.org/docs/apache-airflow/stable/executor/index.html#executor-types

See also #313

### PR DESCRIPTION
**Pull request by adwk67**
# Description

Implements KubernetesExecutor.
Fixes #2.
Jenkins: :green_circle:  https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/airflow-operator-it-custom/99/

Implementation summary:

- the `executor` field has been removed from the top-level cluster config
- the `workers` role has been replaced with a complex enum specifying one of:
  - `celery`: this contains the worker config/roles etc.
  - `kubernetes`: this contains just the logging/resources of the config, plus env-overrides (has to be included as we don't have a role object)
- when `executor = kubernetes`, a pod template is deployed as config map that airflow will use to spawn pods for the DAG tasks 

Tests on Openshift:
```
--- PASS: kuttl (3580.00s)
    --- PASS: kuttl/harness (0.00s)
        --- PASS: kuttl/harness/logging_airflow-2.4.1_openshift-true (282.85s)
        --- PASS: kuttl/harness/ldap_airflow-latest-2.6.1_ldap-authentication-insecure-tls_openshift-true_executor-celery (321.02s)
        --- PASS: kuttl/harness/mount-dags-gitsync_airflow-latest-2.6.1_openshift-true_executor-kubernetes (215.58s)
        --- PASS: kuttl/harness/logging_airflow-2.2.5_openshift-true (305.99s)
        --- PASS: kuttl/harness/logging_airflow-2.2.4_openshift-true (328.94s)
        --- PASS: kuttl/harness/mount-dags-configmap_airflow-latest-2.6.1_openshift-true_executor-kubernetes (202.57s)
        --- PASS: kuttl/harness/mount-dags-configmap_airflow-latest-2.6.1_openshift-true_executor-celery (290.25s)
        --- PASS: kuttl/harness/orphaned-resources_airflow-latest-2.6.1_openshift-true (318.29s)
        --- PASS: kuttl/harness/smoke_airflow-2.2.5_openshift-true_executor-kubernetes (292.73s)
        --- PASS: kuttl/harness/smoke_airflow-2.4.1_openshift-true_executor-kubernetes (267.12s)
        --- PASS: kuttl/harness/smoke_airflow-2.6.1_openshift-true_executor-celery (284.75s)
        --- PASS: kuttl/harness/smoke_airflow-2.4.1_openshift-true_executor-celery (335.22s)
        --- PASS: kuttl/harness/smoke_airflow-2.6.1_openshift-true_executor-kubernetes (226.21s)
        --- PASS: kuttl/harness/cluster-operation_airflow-latest-2.6.1_openshift-true (303.22s)
        --- PASS: kuttl/harness/resources_airflow-latest-2.6.1_openshift-true (209.99s)
        --- PASS: kuttl/harness/smoke_airflow-2.2.4_openshift-true_executor-celery (343.50s)
        --- PASS: kuttl/harness/smoke_airflow-2.2.5_openshift-true_executor-celery (395.81s)
        --- PASS: kuttl/harness/smoke_airflow-2.2.4_openshift-true_executor-kubernetes (314.30s)
        --- PASS: kuttl/harness/logging_airflow-2.6.1_openshift-true (239.35s)
        --- PASS: kuttl/harness/ldap_airflow-latest-2.6.1_ldap-authentication-server-verification-tls_openshift-true_executor-celery (311.55s)
        --- PASS: kuttl/harness/ldap_airflow-latest-2.6.1_ldap-authentication-no-tls_openshift-true_executor-kubernetes (202.59s)
        --- PASS: kuttl/harness/ldap_airflow-latest-2.6.1_ldap-authentication-no-tls_openshift-true_executor-celery (296.19s)
        --- PASS: kuttl/harness/ldap_airflow-latest-2.6.1_ldap-authentication-insecure-tls_openshift-true_executor-kubernetes (246.54s)
        --- PASS: kuttl/harness/mount-dags-gitsync_airflow-latest-2.6.1_openshift-true_executor-celery (251.68s)
        --- PASS: kuttl/harness/ldap_airflow-latest-2.6.1_ldap-authentication-server-verification-tls_openshift-true_executor-kubernetes (200.25s)
```

## Definition of Done Checklist

- Not all of these items are applicable to all PRs, the author should update this template to only leave the boxes in that are relevant
- Please make sure all these things are done and tick the boxes
 
```[tasklist]
# Author
- [x] Changes are OpenShift compatible
- [x] Helm chart can be installed and deployed operator works
- [x] Integration tests passed (add new ones for the executor)
- [x] Documentation
- [x] Logging works for ephemeral worker pods (not working: covered by https://github.com/stackabletech/airflow-operator/issues/318)
- [x] Check tini invocation
- [x] CRD changes are approved
- [x] Ensure vector container is shutdown on the worker pods used by the kubernetes executor. See https://github.com/stackabletech/docker-images/pull/435
```

```[tasklist]
# Reviewer
- [x] Code contains useful comments
- [x] (Integration-)Test cases added
- [x] Documentation added or updated
- [x] Changelog updated
- [x] Cargo.toml only contains references to git tags (not specific commits or branches)
```

```[tasklist]
# Acceptance
- [ ] Feature Tracker has been updated
- [ ] Proper release label has been added
```

### CODE REVIEW

**Comment by adwk67** [comment]
Jenkins after refactoring CRD structure: :green_circle: https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/airflow-operator-it-custom/100/

File: docs/modules/airflow/pages/index.adoc [hunk 1]
```diff
@@ -15,10 +15,47 @@ The Operator manages three https://kubernetes.io/docs/concepts/extend-kubernetes
...
 
 When an AirflowCluster is first deployed, an AirflowDB resource is created. The AirflowDB resource is a wrapper resource for the metadata SQL database that is used by Airflow to store information on users and permissions as well as workflows, task instances and their execution. The resource contains some configuration but also keeps track of whether the database has been initialized or not. It is not deleted automatically if a AirflowCluster is deleted, and so can be reused.
 
+=== Executors
+
+The `worker` role is deployed when `spec.celeryExecutors` is specified (the alternative is `spec.kubernetesExecutors`, whereby pods are created dynamically as needed without jobs being routed through a redis queue to the workers). This means that for `celeryExecutors` there exists an implicit single role which does not appear in resource definition. This is illustrated below:
```
Comment (razvan) [inline docs/modules/airflow/pages/index.adoc:24]:
```suggestion
The `worker` role is deployed when `spec.celeryExecutors` is specified (the alternative is `spec.kubernetesExecutors`, whereby pods are created dynamically as needed without jobs being routed through a redis queue to the workers). This means that for `kubernetesExecutors` there exists an implicit single role which does not appear in resource definition. This is illustrated below:
```

File: docs/modules/airflow/pages/getting_started/first_steps.adoc [hunk 2]
```diff
@@ -56,14 +56,11 @@ Where:
 
 - `metadata.name` contains the name of the Airflow cluster
 - the label of the Docker image provided by Stackable must be set in `spec.version`
-- `spec.executor`: this setting determines how the cluster will run (for more information see https://airflow.apache.org/docs/apache-airflow/stable/executor/index.html#executor-types): the `CeleryExecutor`
-is the recommended setting although `SequentialExecutor` (all jobs run in one process in series) and `LocalExecutor`
-(whereby all jobs are run on one node, using whatever parallelism is possible) are also supported
+- `spec.executor`: this setting determines how the cluster will run (for more information see https://airflow.apache.org/docs/apache-airflow/stable/executor/index.html#executor-types). This field takes one of two values, `celery` or `kubernetes`: `celery` will deploy workers as defined by the role and specify `CeleryExecutor` internally. `kubernetes` will set `KubernetesExecutor` internally, and the scheduler will take responsibility for creating (and terminating) pods for the DAG individual tasks.
```
Comment (razvan) [inline docs/modules/airflow/pages/getting_started/first_steps.adoc:59]:
```suggestion
- `spec.celeryExecutors`: deploy workers managed by Airflow's Celery engine. Alternatively you can use `kuberenetesExectors` tha twill use Airflow's Kubernetes engine for worker management. For more information see https://airflow.apache.org/docs/apache-airflow/stable/executor/index.html#executor-types).
```

**Review by razvan — COMMENTED** [review]
It works and it's ok but IMO:
- `ExecutorConfig` vs `AirflowConfig` - the names are confusing don't reflect the intent.
- Not sure if it's wise to introduce so much complication just because of the affinity issue of executors. But fortunately it's an implementation detail that we can hopefully change later without breaking the CRD.

**Comment by razvan** [comment]
This is almost adding support for complete new product. It may be worth adding an explicit integration test for kubernetes workers and actually run a some tasks with it.

File: tests/templates/kuttl/mount-dags-gitsync/03-assert.yaml.j2 [hunk 3]
```diff
@@ -4,6 +4,7 @@ kind: TestAssert
 metadata:
   name: test-airflow-cluster
 timeout: 1200
+{% if test_scenario['values']['executor'] == 'celery' %}
```
[the remark this answers is not in the document]
Comment (adwk67) [reply tests/templates/kuttl/mount-dags-gitsync/03-assert.yaml.j2:7]: Oops, this was a copy-and-paste error.

**Review by razvan — APPROVED** [review]

**Pull request merged**